CMakeLists.txt

```diff
@@ -0,0 +1,160 @@
+    endif()
+    message(VERBOSE "The C compiler identification is: ${CMAKE_C_COMPILER_ID}. Compiler port folder is: ${FREERTOS_COMPILER_PORT_DIRECTORY}")
+
+    # Search the port folder for available port configurations.
+    file(GLOB PORTS RELATIVE "${FREERTOS_COMPILER_PORT_DIRECTORY}" CONFIGURE_DEPENDS "${FREERTOS_COMPILER_PORT_DIRECTORY}/*")
+    message(VERBOSE "Available ports in ${FREERTOS_COMPILER_PORT_DIRECTORY} are: ${PORTS}")
+
+    # Check that <FREERTOS_KERNEL_PORT> contains a valid port in the identified compiler port directory.
+    if (NOT DEFINED FREERTOS_KERNEL_PORT)
+        message(FATAL_ERROR "<FREERTOS_KERNEL_PORT> is undefined. Set <FREERTOS_KERNEL_PORT> to the desired port. Available ports are: ${PORTS}")
+    else()
+        if(NOT FREERTOS_KERNEL_PORT IN_LIST PORTS)
+            message(FATAL_ERROR "${FREERTOS_KERNEL_PORT} is not an available port. Available ports are: ${PORTS}")
+        else()
+            set(FREERTOS_KERNEL_PORT ${FREERTOS_COMPILER_PORT_DIRECTORY}/${FREERTOS_KERNEL_PORT})
+        endif()
+    endif()
+endif()
+message(VERBOSE "Directory for port implementation is: ${FREERTOS_KERNEL_PORT}")
+
+# Search the port directory for source files.
+file(GLOB FREERTOS_KERNEL_PORT_SOURCES
+    ${FREERTOS_KERNEL_PORT}/*.asm
+    ${FREERTOS_KERNEL_PORT}/*.c
+    ${FREERTOS_KERNEL_PORT}/*.s*
+    ${FREERTOS_KERNEL_PORT}/*.S
+)
+message(VERBOSE "Source files identified for port implementation are: ${FREERTOS_KERNEL_PORT_SOURCES}")
+
+
+## FreeRTOS Kernel Heap ##
+if(FREERTOS_KERNEL_USE_CUSTOM_HEAP)
+    # If a custom heap is being used then check that <FREERTOS_KERNEL_HEAP_FILE> contains a valid file.
+    if(NOT DEFINED FREERTOS_KERNEL_HEAP_FILE)
+        message(FATAL_ERROR "<FREERTOS_KERNEL_HEAP_FILE> is undefined. Set <FREERTOS_KERNEL_HEAP_FILE> to the file that contains the heap implementation.")
+    else()
+        if(NOT EXISTS "${FREERTOS_KERNEL_HEAP_FILE}")
+            message(FATAL_ERROR "Custom heap implementation file: ${FREERTOS_KERNEL_HEAP_FILE} does not exist.")
+        endif()
+    endif()
+else()
+    # Search the memory management directory for available heap implementations.
+    set(FREERTOS_MEMORY_MANAGEMENT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/MemMang")
+    file(GLOB HEAP_FILES RELATIVE "${FREERTOS_MEMORY_MANAGEMENT_DIRECTORY}" CONFIGURE_DEPENDS "${FREERTOS_MEMORY_MANAGEMENT_DIRECTORY}/*.c")
+    message(VERBOSE "Available heap files in ${FREERTOS_MEMORY_MANAGEMENT_DIRECTORY} are: ${HEAP_FILES}")
+
+    # Check that <FREERTOS_KERNEL_HEAP_FILE> contains a valid heap file in the memory management directory.
+    if(NOT DEFINED FREERTOS_KERNEL_HEAP_FILE)
+        message(FATAL_ERROR "<FREERTOS_KERNEL_HEAP_FILE> is undefined. Set <FREERTOS_KERNEL_HEAP_FILE> to the desired heap implementation. Available implementations are: ${HEAPS}")
```

Comment:
The _verbose_ level message doesn't really add value, does it? If the heap file can't be found then the user will get both the verbose and error level messages which state the same thing - the available heap files.

I'd consider removing this (and similar messages in this file). Happy case verbose messages are fine as they won't duplicate.

Reply:
I had it in there as a way to list the available heap implementations without needing to run a misconfigured project for them to print. It would only be duplicated when printing verbose messages with an invalid configuration. I'm fine removing it if that's not useful.